src/main.ts

```diff
@@ -1,63 +1,43 @@
 import { setOutput, setFailed, info } from '@actions/core'
-import add from 'date-fns/add'
-import isAfter from 'date-fns/isAfter'
+import { fold } from 'fp-ts/Either'
+import { pipe } from 'fp-ts/pipeable'
+import * as TE from 'fp-ts/TaskEither'
 import * as inputs from './inputs'
 import { getFeatures } from './optimizely'
+import { Reason, EarlyExit } from './models'
+import { guardLength, toEnabledFeatures, toStaleFeatures } from './tasks'
 
 const run = async () => {
   try {
     info(`getting Features for project "${inputs.PROJECT}"...`)
-    const features = await getFeatures(inputs.PROJECT, inputs.TOKEN)
-    if (features.length === 0) {
-      info('no features found. exiting...')
-      setOutput('features', [])
-      return
-    }
-
-    info(`found ${features.length} features`)
-    info('getting primary environment...')
-    const primaryEnvironment = Object.keys(features[0].environments).reduce<
-      string | null
-    >((acc, key) => {
-      if (acc !== null) {
-        return acc
-      }
-
-      return features[0].environments[key].is_primary ? key : null
-    }, null)
-
-    if (primaryEnvironment === null) {
-      throw new Error(`could not find a primary environment`)
-    }
-
-    info(`primary environment found: "${primaryEnvironment}"`)
-    info('getting enabled features...')
-    const enabledFeatures = features.filter(
-      f =>
-        f.environments[primaryEnvironment].rollout_rules.findIndex(
-          r => r.enabled,
-        ) > -1,
-    )
-    if (enabledFeatures.length === 0) {
-      info('no enabled features. exiting...')
-      setOutput('features', [])
-      return
-    }
-
-    info(`found ${enabledFeatures.length} enabled features`)
-    info(
-      `checking for features enabled for ${inputs.DAYS_UNTIL_STALE} days or more...`,
-    )
-
-    const stale = enabledFeatures.filter(f =>
-      isAfter(
-        new Date(),
-        add(new Date(f.last_modified), { days: inputs.DAYS_UNTIL_STALE }),
-      ),
-    )
-
-    info(`found ${stale.length} stale features`)
-    setOutput('features', stale)
+    pipe(
+      getFeatures(inputs.PROJECT, inputs.TOKEN),
+      TE.chain(guardLength(Reason.NoFeatures)),
+      TE.chain(toEnabledFeatures),
+      TE.chain(guardLength(Reason.NoEnabledFeatures)),
+      TE.chain(toStaleFeatures),
+      TE.chain(guardLength(Reason.NoStaleFeatures)),
+      TE.chain(features => {
```

Comment:
you should use a `TE.bimap` here, which lets you map both left and right in the same function.

you calso dont need to chain this line, if you `return features` instead of creating a new right

Reply:
actually can't you just use a `Either.fold` here?unless you still want the TaskEither

Reply:
I think `fold` is definitely the way to go here, but I'm having a bit of trouble getting it to work. I'll hit you up about this tomorrow.